### ISSUE TITLE
Doesn't work on Windows 10

### ISSUE THREAD

**Issue by enjoysmath** [enlyze/PortSniffer#7]
The below says to restart my machine.  That doesn't work to fix the problem.   It simply kills my COM ports, and doesn't monitor anything.

Luckily I can get back my COM ports by unininstalling the device(s) in the Device Manager.

```
c:\Users\fruit\OneDrive\Desktop\SoundUnited\WindowsDriverDev\PortSniffer\redist_AMD64>PortSniffer-Tool.exe /ports
**********************************************************************
ENLYZE PortSniffer Tool 1.3 (unknown revision)
Copyright 2020-2021 Colin Finck, ENLYZE GmbH <c.finck@enlyze.com>
**********************************************************************

COM1
COM4

c:\Users\fruit\OneDrive\Desktop\SoundUnited\WindowsDriverDev\PortSniffer\redist_AMD64>PortSniffer-Tool.exe /attach COM4
**********************************************************************
ENLYZE PortSniffer Tool 1.3 (unknown revision)
Copyright 2020-2021 Colin Finck, ENLYZE GmbH <c.finck@enlyze.com>
**********************************************************************

The PortSniffer Driver has been successfully attached to COM4.
A reboot is required for the changes to take effect.

c:\Users\fruit\OneDrive\Desktop\SoundUnited\WindowsDriverDev\PortSniffer\redist_AMD64>PortSniffer-Tool.exe /monitor COM4 R
**********************************************************************
ENLYZE PortSniffer Tool 1.3 (unknown revision)
Copyright 2020-2021 Colin Finck, ENLYZE GmbH <c.finck@enlyze.com>
**********************************************************************

Could not open "\\.\EnlyzePortSniffer", last error is 2.
Either the PortSniffer Driver is not installed or not attached to any port.
```

Can this be made to work for Windows 10?

**Comment by ColinFinck** [comment]
Hi! I've just updated the relevant chapter in the README file: https://github.com/enlyze/PortSniffer#driver-signature
Long story short: The current PortSniffer driver is not signed, so you temporarily need to disable driver signature enforcement at boot to work around this Microsoft limitation. The README explains this in detail now.

**Comment by enjoysmath** [comment]
@ColinFinck  is there any way to disable driver sig enforcement programmatically?  

I will need to put in detailed instructions with images for my users, but that's okay for now.   Could I get a driver signed just for internal use of my engineering team?

**Comment by enjoysmath** [comment]
Thank you, Colin!

You've been very helpful and responsive.

I/you got it to work!

**Comment by enjoysmath** [comment]
@ColinFinck  Here are the Windows 10 driver signing requirements:

https://learn.microsoft.com/en-us/windows-hardware/drivers/install/kernel-mode-code-signing-requirements--windows-vista-and-later-

Precisely what requirements are not met by your driver?

**Comment by ColinFinck** [comment]
> @ColinFinck is there any way to disable driver sig enforcement programmatically?

I'm not aware of any solution other than disabling it manually for every boot.

> Could I get a driver signed just for internal use of my engineering team?

I didn't need a certificate for my usage scenarios yet, so I don't have any I could use.
As per [latest Microsoft guidelines on that topic](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/kernel-mode-code-signing-policy--windows-vista-and-later-), it now also needs to be a costly EV Code Signing certificate if you want to install the driver on latest Windows without any hassles.

At least, I found [this guide about dual-signing](https://kerneldrivers.com/dual-signed-binaries-windows-7-beyond/) that puts some hope into the ability to ship a single signed driver for Windows 7 and any later version. The situation doesn't seem to be [as bad as OSR puts it](https://www.osr.com/blog/2021/04/08/lost-cause-no-driver-updates-allowed-except-for-win-10/) where Attestation Signing is described as an option only for Windows 10+ drivers.
Dual-signing with a cross-certificate and Attestation Signing could be worth a look if I ultimately decide to get an EV Code Signing certificate.
What still doesn't look like an option to me is pushing my driver through all stages of the Hardware Lab Kit (HLK). I don't have the capacity to set up a hardware lab of multiple machines, with at least one of them required to be a costly Windows Server SKU, just to test my ~1000 lines driver and get an official blessing.

In any case, you are free to sign the driver yourself with a test certificate. This is described [here](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/creating-test-certificates).
I haven't tested it myself though, so I can't say whether it "just works" after installing your certificate as a CA in the system or whether it still needs the "Test Mode" option enabled at boot. In the worst case, it may not be better at all than the current unsigned driver.

**Comment by enjoysmath** [comment]
@ColinFinck Thanks for all the information on that. I may do the Test Mode you mentioned at the bottom if my coding bosses say that they're uncomfortable with the Driver Signature Enforcement disabling.  I made an installer program in PyQt5 that guides the user through the steps.  So hopefully that makes it more "user friendly". :)

I wanted to thank you for making such an amazing driver / tool for us to use.  I think Enlyze/PortSniffer is the only option available for (programatically) sniffing traffic on the serial ports on Windows.    I'm making an test tool for automated test procedures using Murideo 7G hardware and their controller app as what I'm recording.   There will also be some ADB buttons to press.  The engineer hits record and I record the byte stream between the controller app and the Murideo hardware.  The hardware in turn runs tests on the test setup which includes a monitor and a certain consume electronics company's 
audio device.

**Comment by enjoysmath** [comment]
Colin, I sent you two emails.  I'm now getting:

```
c:\Users\fruit\OneDrive\Desktop\SoundUnited\SoundUnitedEngineersTestTool>cd output

c:\Users\fruit\OneDrive\Desktop\SoundUnited\SoundUnitedEngineersTestTool\output>PortSniffer-Tool.exe /install
**********************************************************************
ENLYZE PortSniffer Tool 2.0 (ee05689b3976a537dea882d2d9994b51399f338f)
Copyright 2020-2022 Colin Finck, ENLYZE GmbH <c.finck@enlyze.com>
**********************************************************************

CopyFileW failed, last error is 32.

c:\Users\fruit\OneDrive\Desktop\SoundUnited\SoundUnitedEngineersTestTool\output>

```

for both version 1.4 and 2.0.  I've even built the driver on my machine and the same issue.  Before I looked into this issue, it was killing my ports (again).  Is there some kind of protocol you have to follow using this, such as always /detaching after I'm done?  

Thanks.

**Comment by enjoysmath** [comment]
My fault (I guess) I didn't know you have to disable DSE every boot.  That sucks.  I will let my users know and I'll try out the Test Signing thing.

**Comment by enjoysmath** [comment]
I have it interfaced now with my PyQt5 app, and it's working great.  Let me know (anyone) if you'd like to know how I got it to work.  Can't do it simply with subprocess.Popen.  Essentially, had to use WinPty library for Python and a regex specific to the data bytes that I'm parsing.

**Comment by ColinFinck** [comment]
@enjoysmath Thanks for all your feedback!

I did some more testing in the last few days and could reproduce some of the problems you were experiencing.
My just released PortSniffer 2.1 fixes them and is much more robust.

In particular, you got `CopyFileW failed, last error is 32`, because starting with Windows 10, the operating system locks driver files in use. In the case of PortSniffer, it doesn't release them until the next reboot, even when PortSniffer has already been uninstalled.
As a consequence, uninstallation previously failed with a `DeleteFileW` error on Windows 10, and a subsequent installation of a different PortSniffer version failed with the `CopyFileW` error above.
Sadly, this is only fixable by requiring a reboot after every uninstallation. Which is exactly what PortSniffer 2.1 does now.
Alas, this also means that upgrading from one PortSniffer version to the other requires a clean uninstallation of the previous version and a reboot. But I'm better safe than sorry here!

This also explains why you couldn't get PortSniffer 2.0 to work, but only 1.4.
When not rebooting between upgrades, the old PortSniffer driver stays in memory. As a consequence, you were running the PortSniffer 2.0 tool with the incompatible PortSniffer 1.4 driver, causing errors such as `Captured an invalid request type: 0x0000`.
To prevent such problems in the future, `PortSniffer-Tool /monitor` now checks the driver and tool versions before doing anything, and bails out if they are incompatible.

Finally, I checked why you couldn't get any output from PortSniffer-Tool using `subprocess.Popen` and instead required `winpty`.
It turned out that PortSniffer-Tool _with a redirected stdout_ was buffering all text output for a potentially long time. This only happened with redirected output and not in an interactive terminal environment.
In any case, I have now disabled that buffering in PortSniffer 2.1. You should now be able to capture the tool's output using `subprocess.Popen` or any other stdout redirector. A detour via `winpty` should no longer be necessary.
